Modules/_abc.c

```diff
@@ -0,0 +1,826 @@
+    }
+    if (result < 0) {
+        return NULL;
+    }
+    /* Subtle: test for cycles *after* testing for "already a subclass";
+       this means we allow X.register(X) and interpret it as a no-op. */
+    result = PyObject_IsSubclass(self, subclass);
+    if (result > 0) {
+        /* This would create a cycle, which is bad for the algorithm below. */
+        PyErr_SetString(PyExc_RuntimeError, "Refusing to create an inheritance cycle");
+        return NULL;
+    }
+    if (result < 0) {
+        return NULL;
+    }
+    _abc_data *impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+    if (_add_to_weak_set(impl->_abc_registry, subclass) < 0) {
+        Py_DECREF(impl);
+        return NULL;
+    }
+    Py_DECREF(impl);
+
+    /* Invalidate negative cache */
+    PyObject *one = PyLong_FromLong(1);
+    if (one == NULL) {
+        return NULL;
+    }
+    PyObject *next_version = PyNumber_Add(abc_invalidation_counter, one);
+    Py_DECREF(one);
+    if (next_version == NULL) {
+        return NULL;
+    }
+    Py_SETREF(abc_invalidation_counter, next_version);
+
+    Py_INCREF(subclass);
+    return subclass;
+}
+
+PyDoc_STRVAR(_abc_instancecheck_doc,
+"Internal ABC helper for instance checks. Should be never used outside abc module");
+
+static PyObject *
+_abc_instancecheck(PyObject *m, PyObject *args)
+{
+    PyObject *self, *result = NULL, *subclass = NULL,
+             *subtype, *instance = NULL;
+    if (!PyArg_UnpackTuple(args, "_abc_instancecheck", 2, 2, &self, &instance)) {
+        return NULL;
+    }
+
+    _abc_data *impl = _get_impl(self);
+    if (impl == NULL) {
+        return NULL;
+    }
+
+    subclass = _PyObject_GetAttrId(instance, &PyId___class__);
+    /* Inline the cache checking. */
+    int incache = _in_weak_set(impl->_abc_cache, subclass);
+    if (incache < 0) {
+        goto end;
+    }
+    if (incache > 0) {
+        result = Py_True;
+        Py_INCREF(result);
+        goto end;
+    }
+    subtype = (PyObject *)Py_TYPE(instance);
+    if (subtype == subclass) {
+        int r = PyObject_RichCompareBool(
+            impl->_abc_negative_cache_version,
+            abc_invalidation_counter, Py_EQ);
+        assert(r >= 0);  // Both should be PyLong
+        if (r > 0) {
+            incache = _in_weak_set(impl->_abc_negative_cache, subclass);
+            if (incache < 0) {
+                goto end;
+            }
+            if (incache > 0) {
+                result = Py_False;
+                Py_INCREF(result);
+                goto end;
+            }
+        }
+        /* Fall back to the subclass check. */
+        result = _PyObject_CallMethodIdObjArgs(self, &PyId___subclasscheck__,
+                                               subclass, NULL);
+        goto end;
+    }
+    result = _PyObject_CallMethodIdObjArgs(self, &PyId___subclasscheck__,
+                                           subclass, NULL);
+    if (result == NULL || result == Py_True) {
```

Comment:
Deviation in behavior for `__subclasscheck__` methods that return truthy non-bools, which would get coerced to false in this case.